### PR TITLE
Validate migration files

### PR DESCRIPTION
**Pull request by realaravinth**
On behalf of [@dachary](https://lab.forgefriends.org/dachary) from the [forgefriends](https://forgefriends.org) project:

JSON Schema validation for data used by Gitea during migrations
Discussion at https://forum.forgefriends.org/t/common-json-schema-for-repository-information/563

---
[source](https://lab.forgefriends.org/forgefriends/forgefriends/-/merge_requests/32)

### CODE REVIEW

**Comment by lunny** [comment]
I think it's unnecessary to increment the binary size for a test purpose.

**Comment by realaravinth** [comment]
> I think it's unnecessary to increment the binary size for a test purpose.

The bindata keyword that was added to .drone.yml and Makefile was removed. Thanks for the quick review :-)

---
[source](https://lab.fedeproxy.eu/forgefriends/forgefriends/-/merge_requests/32#note_763)

**Review by lunny — APPROVED** [review]

**Comment by realaravinth** [comment]
on behalf of [@dachary](https://lab.forgefriends.org/dachary):

Thanks for the review. The bug fix was incorrect. The milestone fixtures must have a `deadline_unix` field set instead. I now understand why the year 9999 is special.

---
[source](https://lab.fedeproxy.eu/forgefriends/forgefriends/-/merge_requests/33#note_794)

**Comment by realaravinth** [comment]
The PR was updated to remove the conflicting changes in the vendor directory.

---
[source](https://lab.fedeproxy.eu/forgefriends/forgefriends/-/merge_requests/32#note_818)

File: modules/migration/file_format.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migration
+
+import (
+	"encoding/json"
```
Comment (Gusted) [inline modules/migration/file_format.go]: Sigh, we really need to get a linter for this specific code.

`encoding/json` should be replaced by `code.gitea.io/gitea/modules/json`

File: modules/migration/file_format.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+	"github.com/santhosh-tekuri/jsonschema/v5"
+	"gopkg.in/yaml.v2"
+)
+
+// Load project data from file, with optional validation
+func Load(filename string, data interface{}, validation bool) error {
+	isJSON := strings.HasSuffix(filename, ".json")
+
+	_, err := os.Stat(filename)
```
Comment (Gusted) [inline modules/migration/file_format.go]: Why is stat necessary? If this is purely to check if the file exists - `ReadFile` should give a error for that as well.

File: modules/migration/file_format.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+		if err != nil {
+			return err
+		}
+		err = sch.Validate(v)
+		if err != nil {
+			log.Error("migration validation failed for\n%s", string(bs))
+		}
+		return err
+	default:
+		return fmt.Errorf("file_format:validate: %T is not a known time", datatype)
```
Comment (Gusted) [inline modules/migration/file_format.go]:
```suggestion
		return fmt.Errorf("file_format:validate: %T has not a validation implemented", datatype)
```


`Is not a known time` sounds rather confusing to me if I got that error as developer.

File: modules/migration/schemas/issue.json [hunk 4]
```diff
@@ -0,0 +1,114 @@
+{
+    "title": "Issue",
+    "description": "Issues associated to a repository within a forge.",
```
Comment (Gusted) [inline modules/migration/schemas/issue.json]: Off-topic: what is the definition of `forge` in this specific context?

File: modules/migration/schemas/issue.json [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+	    "state",
+	    "is_locked",
+	    "created",
+	    "updated"
+	]
+    },
+
+    "$schema": "http://json-schema.org/draft-04/schema#",
+    "$id": "http://example.com/issue.json",
+    "$$target": "issue.json"
```
Comment (Gusted) [inline modules/migration/schemas/issue.json:113]: I had a look around the jsonschema library, but it doesn't seem that `$$target` is used anywhere?

**Review by Gusted — CHANGES_REQUESTED** [review]

**Comment by 6543** [comment]
CI fail related:
```
+ make lint-backend
golangci-lint run --timeout 10m
services/migrations/dump.go:155:3: not enough arguments in call to git.Clone (typecheck)
	})
	 ^
services/migrations/dump.go:162:56: not enough arguments in call to repository.WikiRemoteURL (typecheck)
		wikiRemotePath := repository.WikiRemoteURL(remoteAddr)
		                                                     ^
services/migrations/dump.go:173:5: not enough arguments in call to git.Clone (typecheck)
			}); err != nil {
			 ^
make: *** [Makefile:754: golangci-lint] Error 1
```

File: modules/migration/file_format.go (hunk 1 above)
Comment (realaravinth) [reply modules/migration/file_format.go]: Fixed.

File: modules/migration/file_format.go (hunk 2 above)
Comment (realaravinth) [reply modules/migration/file_format.go]: You are correct, removed.

File: modules/migration/file_format.go (hunk 3 above)
Comment (realaravinth) [reply modules/migration/file_format.go]: Your improved error message was used instead.

File: modules/migration/schemas/issue.json (hunk 4 above)
Comment (realaravinth) [reply modules/migration/schemas/issue.json]: Examples were added for clarification.

File: modules/migration/schemas/issue.json (hunk 5 above)
Comment (realaravinth) [reply modules/migration/schemas/issue.json:113]: Not yet indeed, but it will be at a later time.

**Comment by realaravinth** [comment]
@Gusted thanks for your review. I addressed your comments, hopefully to your satisfaction

---
[source](https://lab.fedeproxy.eu/forgefriends/forgefriends/-/merge_requests/32#note_880)

**Comment by realaravinth** [comment]
@6543 thanks for highlighting the CI failure. The PR was rebased against today's main branch to resolve them.
---
[source](https://lab.fedeproxy.eu/forgefriends/forgefriends/-/merge_requests/32#note_881)

File: modules/migration/schemas_bindata.go [hunk 6]
```diff
@@ -0,0 +1,10 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+//go:build bindata
```
Comment (Gusted) [inline modules/migration/schemas_bindata.go:5]: Could this filename be changed bindata -> static

**Review by Gusted — COMMENTED** [review]
Only a small nit, otherwise it looks good to me.

File: modules/migration/schemas_bindata.go (hunk 6 above)
Comment (realaravinth) [reply modules/migration/schemas_bindata.go:5]: It is meant to be consistent with the naming of all other bindata files ( BINDATA_DEST := modules/public/bindata.go modules/options/bindata.go modules/templates/bindata.go ) generators ( modules/public/public_bindata.go modules/options/options_bindata.go modules/templates/templates_bindata.go ) and not to be confused with the name of the file that contains the code modules/migration/schemas_static.go.

If that was already clear to you, I may have misunderstood your suggestion?

---
[source](https://lab.fedeproxy.eu/forgefriends/forgefriends/-/merge_requests/32#note_885)
[typo fix source](https://lab.fedeproxy.eu/forgefriends/forgefriends/-/merge_requests/32#note_886)

Comment (Gusted) [reply modules/migration/schemas_bindata.go:5]: > If that was already clear to you, I may have misunderstood your suggestion?

It was not, I think I might have misunderstood another PR that was using this format, sorry!

**Review by Gusted — APPROVED** [review]
LGTM!

File: modules/migration/milestone.go [hunk 7]
```diff
@@ -9,11 +9,11 @@ import "time"
...
-	Updated     *time.Time
-	Closed      *time.Time
-	State       string // open, closed
+	Title       string     `json:"title"`
+	Description string     `json:"description"`
+	Deadline    *time.Time `json:"deadline"`
+	Created     time.Time  `json:"created"`
+	Updated     *time.Time `json:"updated"`
+	Closed      *time.Time `json:"closed"`
+	State       string     `json:"state"`
```
Comment (lunny) [inline modules/migration/milestone.go]: Ah, the comment should be kept.

Comment (realaravinth) [reply modules/migration/milestone.go]: @lunny good catch! The comment was restored.

---
[source](https://lab.fedeproxy.eu/forgefriends/forgefriends/-/merge_requests/32#note_889)

**Comment by 6543** [comment]
please dont force-push I cant easily check what  was changed

**Review by 6543 — CHANGES_REQUESTED** [review]
force push after 2 lgtms - need :eyes:

**Comment by realaravinth** [comment]
@6543: Apologies, we'll avoid force pushing next time

**Comment by 6543** [comment]
well we squash-merge so its better to not force-push to see the progress

**Review by 6543 — APPROVED** [review]

**Pull request merged**